lib/Dialect/RTL/Ops.cpp

```diff
@@ -657,12 +657,72 @@ void AddOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
         return success();
       }
 
+      // add(..., x, x) -> add(..., shl(x, 1))
+      if (inputs[size - 1] == inputs[size - 2]) {
+        SmallVector<Value, 4> newOperands(inputs.drop_back(/*n=*/2));
+
+        auto one = rewriter.create<ConstantOp>(
+            op.getLoc(), 1, op.getType().cast<IntegerType>());
+        auto shiftLeftOp =
+            rewriter.create<rtl::ShlOp>(op.getLoc(), inputs.back(), one);
+
+        newOperands.push_back(shiftLeftOp);
+        rewriter.replaceOpWithNewOp<AddOp>(op, op.getType(), newOperands);
+        return success();
+      }
+
+      auto shlOp = inputs[size - 1].getDefiningOp<rtl::ShlOp>();
+      APInt shift;
+
+      // add(..., x, shl(x, c)) -> add(..., mul(x, c * 2 + 1))
+      if (shlOp && shlOp.lhs() == inputs[size - 2] &&
+          matchPattern(shlOp.rhs(), m_RConstant(shift))) {
+
+        auto numBits =
+            shlOp.getOperand(1).getType().cast<IntegerType>().getWidth();
+        APInt one(numBits, 1, /*isSigned=*/false);
+        APInt two(numBits, 2, /*isSigned=*/false);
+
+        auto multiplier = rewriter.create<ConstantOp>(
+            op.getLoc(), (shift * two + one).getLimitedValue(),
+            op.getType().cast<IntegerType>());
+
+        SmallVector<Value, 2> mulInputs;
+        mulInputs.push_back(shlOp.lhs());
+        mulInputs.push_back(multiplier);
+        auto mulOp = rewriter.create<rtl::MulOp>(op.getLoc(), mulInputs);
+
+        SmallVector<Value, 4> newOperands(inputs.drop_back(/*n=*/2));
+        newOperands.push_back(mulOp);
+        rewriter.replaceOpWithNewOp<AddOp>(op, op.getType(), newOperands);
+        return success();
+      }
+
+      auto mulOp = inputs[size - 1].getDefiningOp<rtl::MulOp>();
+      APInt multiplier;
+
+      // add(..., x, mul(x, c) -> add(..., shl(x, s)), where (c + 1) is a
```

Comment:
Could you simplify this to just being the `add(... x, mul(x, c))` -> `add(... mul(x, c+1))` pattern, and let the mul by power of two recursively simplify?

Also, I think this needs to check to see if the mulOp has two operands before swallowing it.

Reply:
> Also, I think this needs to check to see if the mulOp has two operands before swallowing it.
Noted. I'll fix this and make a corresponding test if this is the case, which I believe you're right.

Can you explain what you mean by 
> let the mul by power of two recursively simplify?
I'm not sure I'm following. How will it recursively simplify if we don't have a transition from `mul` -> `shl`?

Reply:
You can write this to just naively create a new "mul x, c".  When c is a power of two, a separate (already existing?) canonicalization can turn "mul x, 4" into "shl x, 2", so you don't have to handle the power of two case with specific code here.

Reply:
So I removed any mention of `shl` in add. Instead, add will recursively commit to `mul(x, c) -> mul(x, c + 1)`. Then, in `mul`, there's a pass for when `inputs.size() == 2` (to avoid early conflicts with constant folds) to turn a `mul(x, c) -> mul(shl(x, n))`. Let me know your thoughts when you have a chance.

Reply:
sounds great!